### PR TITLE
[Merged by Bors] - feat: SMul instance for AddSubmonoid and lemmas

### PR DESCRIPTION
**Pull request by alreadydone**
+ Introduce a `SMul` instance for two `AddSubmonoid`s and provide lemmas

+ Introduce a `Mul` instance for `AddSubgroup`

+ Prove lemmas for the existing `Mul` instance on `AddSubmonoid`, which are generalized from corresponding `Submodule` lemmas

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Submonoid/Pointwise.lean [hunk 1]
```diff
@@ -588,6 +615,33 @@ theorem pow_subset_pow {s : AddSubmonoid R} {n : ℕ} : (↑s : Set R) ^ n ⊆ 
 
 end Semiring
 
+section SMul
+
+variable [AddMonoid R] [DistribSMul R A]
+
+instance : SMul (AddSubmonoid R) (AddSubmonoid A) where
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Submonoid/Pointwise.lean:622]: Is `A := AddSubmonoid R` possible here? If so, then this instance probably creates diamonds

File: Mathlib/Algebra/Group/Subgroup/Pointwise.lean [hunk 2]
```diff
@@ -547,4 +547,19 @@ theorem le_pointwise_smul_iff₀ {a : α} (ha : a ≠ 0) {S T : AddSubgroup A} :
...
 
+section Mul
+
+variable {R : Type*} [NonUnitalNonAssocRing R]
+
+/-- For additive subgroups `S` and `T` of a ring, the product of `S` and `T` as submonoids
+is automatically a subgroup, which we define as the product of `S` and `T` as subgroups. -/
+protected def mul : Mul (AddSubgroup R) where
+  mul M N :=
+  { __ := M.toAddSubmonoid * N.toAddSubmonoid
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Subgroup/Pointwise.lean:558]: Can you add the lemma `(M * N).toAddSubmonoid = M.toAddSubmonoid * N.toAddSubmonoid`?

File: Mathlib/Algebra/Group/Submonoid/Pointwise.lean (hunk 1 above)
Comment (alreadydone) [reply Mathlib/Algebra/Group/Submonoid/Pointwise.lean:622]: For `Semiring R` it's possible, but there's no diamond:
```
example {R} [Semiring R] : Mul.toSMul (AddSubmonoid R) = AddSubmonoid.smul := rfl
```
For NonUnital/NonAssocSemiring there's strangely no DistribSMul instance. For Semiring it goes through [DistribMulAction.toDistribSMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Action/Defs.html#DistribMulAction.toDistribSMul) but for NonUnital/NonAssoc there's no MulAction.

I should definitely change this def to `protected def` and `scoped[Pointwise] attribute [instance]` like the others though ...

Comment (alreadydone) [reply Mathlib/Algebra/Group/Submonoid/Pointwise.lean]: Oh sorry I misread, you're talking about `SMul (AddSubmonoid R) (AddSubmonoid (AddSubmonoid R))` ...

Comment (alreadydone) [reply Mathlib/Algebra/Group/Submonoid/Pointwise.lean]: There's no existing `SMul (AddSubmonoid R) (AddSubmonoid (AddSubmonoid R))` instance, even for [CommRing R].
```
import Mathlib
open Pointwise
variable {R} [CommRing R]
#synth SMul (AddSubmonoid R) (AddSubmonoid <| AddSubmonoid R)
-- failed to synthesize AddZeroClass (AddSubmonoid R)
```
The reason is that there aren't many instances declared on [AddSubmonoid](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Submonoid/Defs.html#AddSubmonoid). There's much more instances on Submodule, e.g. [Submodule.idemSemiring](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Operations.html#Submodule.idemSemiring), so we should probably worry more about that ...

**Review by jcommelin — COMMENTED** [review]
bors d+

File: Mathlib/Algebra/Group/Submonoid/Pointwise.lean [hunk 3]
```diff
@@ -588,6 +615,38 @@ theorem pow_subset_pow {s : AddSubmonoid R} {n : ℕ} : (↑s : Set R) ^ n ⊆ 
...
 
+section SMul
+
+variable [AddMonoid R] [DistribSMul R A]
+
+/-- For `M : Submonoid R` and `N : AddSubmonoid A`, `M • N` is the additive submonoid
+generated by all `m • n` where `m ∈ M` and `n ∈ N`. -/
+protected def smul : SMul (AddSubmonoid R) (AddSubmonoid A) where
+  smul M N := ⨆ s : M, N.map (DistribSMul.toAddMonoidHom A s.1)
+scoped[Pointwise] attribute [instance] AddSubmonoid.smul
```
[the remark this answers is not in the document]
Comment (alreadydone) [reply Mathlib/Algebra/Group/Submonoid/Pointwise.lean:627]: The majority of this file doesn't insert a blank line between the def and the scoped instance declaration, but in the other file a blank line is always inserted. I think both are acceptable but a blank line seems to conform better to the usual style.

**Review by alreadydone — COMMENTED** [review]
Thanks!
bors r+

**Pull request closed without merging**